### PR TITLE
gh-93421: Update sqlite3 cursor.rowcount after SQLITE_DONE

### LINKED ISSUE #93421: SQLite rowcount is corrupted when combining UPDATE RETURNING w/ table that is dropped and recreated
version info:

```
$ python
Python 3.10.0 (default, Nov  5 2021, 17:23:47) [GCC 11.2.1 20210728 (Red Hat 11.2.1-1)] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import sqlite3
>>> sqlite3.sqlite_version
'3.36.0'
```

we have a test suite that creates a table, runs some SQL, then drops it.   if multiple tests run that each perform this task, if the same SQLite connection is used, rowcount starts returning "0".   Seems to also require RETURNING to be used.   Full demonstration:

```py
import os
import sqlite3



def go():
    """function creates a new table, runs INSERT/UPDATE, drops table,
    commits connection.

    """

    # create table
    cursor = conn.cursor()
    cursor.execute(
        """CREATE TABLE some_table (
        id INTEGER NOT NULL,
        value VARCHAR(40) NOT NULL,
        PRIMARY KEY (id)
    )
    """
    )
    cursor.close()
    conn.commit()

    # run operation
    cursor = conn.cursor()
    cursor.execute(
        "INSERT INTO some_table (id, value) VALUES (1, 'v1')"
    )
    ident = 1

    cursor.execute(
        "UPDATE some_table SET value='v2' "
        "WHERE id=? RETURNING id",
        (ident,),
    )
    new_ident = cursor.fetchone()[0]
    assert ident == new_ident
    assert cursor.rowcount == 1, cursor.rowcount
    cursor.close()

    # drop table
    cursor = conn.cursor()
    cursor.execute("DROP TABLE some_table")
    cursor.close()

    conn.commit()

if os.path.exists("file.db"):
    os.unlink("file.db")

# passes
conn = sqlite3.connect("file.db")
go()

# run again w/ new connection (same DB), passes
conn = sqlite3.connect("file.db")
go()

print("FAILURE NOW OCCURS")
# run again w/ same connection, fails
go()
```

on the third run, where we ran the "test" on the same connection twice, it fails:

```
$  python test3.py 
FAILURE NOW OCCURS
Traceback (most recent call last):
  File "/home/classic/dev/sqlalchemy/test3.py", line 62, in <module>
    go()
  File "/home/classic/dev/sqlalchemy/test3.py", line 39, in go
    assert cursor.rowcount == 1, cursor.rowcount
AssertionError: 0
```

it would appear there's some internal caching of table state that needs to be cleared.

### PR DESCRIPTION
**Pull request by erlend-aasland**
Alternative to #93520 
Resolves #93421

### CODE REVIEW

**Comment by erlend-aasland** [comment]
FYI, the failing buildbot are also failing on `main`; they are unrelated to this PR.

**Comment by erlend-aasland** [comment]
I'm going to merge this in a day or two.

**Comment by animalize** [comment]
`rowcount`'s type should also changed.
Otherwise on 32-bit big-endian platform, it will alway 0.
```diff
-{"rowcount", T_LONG, offsetof(pysqlite_Cursor, rowcount), READONLY},
+{"rowcount", T_LONGLONG, offsetof(pysqlite_Cursor, rowcount), READONLY},
```

File: Modules/_sqlite/cursor.c [hunk 1]
```diff
@@ -867,6 +876,7 @@ _pysqlite_query_execute(pysqlite_Cursor* self, int multiple, PyObject* operation
 
     stmt_reset(self->statement);
     stmt_mark_dirty(self->statement);
+    self->rowcount = self->statement->is_dml ? 0L : -1L;
```
Comment (ghost) [inline Modules/_sqlite/cursor.c:869]: `0LL`, `-1LL`.
This makes it easy for readers to understand the type of `rowcount`.

**Comment by animalize** [comment]
These logical branches can be refactored, at least for 3.12 branch.
```c
if (multiple) {
    ...
} else {
   ...
}
```
```c
        if (self->statement->is_dml && rc == SQLITE_DONE && multiple) {
            self->rowcount += changes(self->connection->db);
        }

        if (rc == SQLITE_DONE && !multiple) {
            if (self->statement->is_dml) {
                self->rowcount = changes(self->connection->db);
            }
            stmt_reset(self->statement);
            Py_CLEAR(self->statement);
        }
        if (multiple) {
            stmt_reset(self->statement);
        }
        Py_XDECREF(parameters);
    }
    if (!multiple) {
        sqlite_int64 lastrowid;
        Py_BEGIN_ALLOW_THREADS
        lastrowid = sqlite3_last_insert_rowid(self->connection->db);
        Py_END_ALLOW_THREADS
        Py_SETREF(self->lastrowid, PyLong_FromLongLong(lastrowid));
        // Fall through on error.
    }
```

**Comment by erlend-aasland** [comment]
> `rowcount`'s type should also be changed. Otherwise on 32-bit big-endian platform, it will always be `0`.

Good catch! `long long` is guaranteed to be at least 64 bits.

> How about only apply 64-bit function to 3.12 branch? It's more or less a new feature.

I'm fine with backporting that, but I'll remove it from this PR nonetheless. It makes the PR cleaner.

> These logical branches can be refactored for 3.12 branch.

Yes, we could do a lot of refactoring in the whole sqlite3 module. I prefer to keep refactorings in separate PRs.

Thanks for reviewing!

**Comment by erlend-aasland** [comment]
FTR, buildbot failures are unrelated to this PR.

**Pull request merged**